### PR TITLE
Support `--print KIND=PATH` command line syntax

### PR DESCRIPTION
**Pull request by dtolnay**
As is already done for `--emit KIND=PATH` and `-L KIND=PATH`.

In the discussion of #110785, it was pointed out that `--print KIND=PATH` is nicer than trying to apply the single global `-o` path to `--print`'s output, because in general there can be multiple print requests within a single rustc invocation, and anyway `-o` would already be used for a different meaning in the case of `link-args` and `native-static-libs`.

I am interested in using `--print cfg=PATH` in Buck2. Currently Buck2 works around the lack of support for `--print KIND=PATH` by [indirecting through a Python wrapper script](https://github.com/facebook/buck2/blob/d43cf3a51a31f00be2c2248e78271b0fef0452b4/prelude/rust/tools/get_rustc_cfg.py) to redirect rustc's stdout into the location dictated by the build system.

From skimming Cargo's usages of `--print`, it definitely seems like it would benefit from `--print KIND=PATH` too. Currently it is working around the lack of this by inserting `--crate-name=___ --print=crate-name` so that it can look for a line containing `___` as a delimiter between the 2 other `--print` informations it actually cares about. This is commented as a "HACK" and "abuse". https://github.com/rust-lang/cargo/blob/31eda6f7c360d9911f853b3014e057db61238f3e/src/cargo/core/compiler/build_context/target_info.rs#L242 (FYI @weihanglo as you dealt with this recently in https://github.com/rust-lang/cargo/pull/11633.)

Mentioning reviewers active in #110785: @fee1-dead @jyn514 @bjorn3

### CODE REVIEW

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 1]
```diff
@@ -1459,11 +1465,17 @@ fn print_native_static_libs(
...
         }
     }
-    if !lib_args.is_empty() {
-        sess.emit_note(errors::StaticLibraryNativeArtifacts);
-        // Prefix for greppability
-        // Note: This must not be translated as tools are allowed to depend on this exact string.
-        sess.note_without_error(format!("native-static-libs: {}", &lib_args.join(" ")));
+
+    match out {
+        OutFileName::Real(_) => out.overwrite(&lib_args.join(" "), sess),
```
Comment (b-naber) [inline compiler/rustc_codegen_ssa/src/back/link.rs:1470]: I feel that just writing the arguments here could potentially be confusing without the output that `emit_note(errors::StaticLibraryNativeArtifacts)` would print to stdout. Can we either include that note or maybe don't write anything here at all?

File: compiler/rustc_driver_impl/src/lib.rs [hunk 2]
```diff
@@ -761,15 +778,15 @@ fn print_crate_info(
...
-                if *req == PrintRequest::CrateName {
-                    safe_println!("{id}");
-                    continue;
-                }
-                let crate_types = collect_crate_types(sess, attrs);
-                for &style in &crate_types {
-                    let fname =
-                        rustc_session::output::filename_for_input(sess, style, id, &t_outputs);
-                    safe_println!("{}", fname.as_path().file_name().unwrap().to_string_lossy());
+                if req.kind == CrateName {
```
Comment (b-naber) [inline compiler/rustc_driver_impl/src/lib.rs:781]: Even though this PR didn't introduce this, can you maybe create two separate match arms for `FileNames` and `CrateName`? This does introduce a bit of code duplication, but we wouldn't have the `build_output_filenames` in the `CrateName` arm and this seems a little cleaner overall imo.

File: compiler/rustc_codegen_ssa/src/back/link.rs (hunk 1 above)
Comment (dtolnay) [reply compiler/rustc_codegen_ssa/src/back/link.rs:1470]: Good call. How about keeping a `sess.emit_note` that informs you that it wrote the native-static-libs to a file? Similar to what #76843 did for long type names -- write the unadorned information to a file (which is the most useful for tooling) and also write an informative note among the compiler diagnostics to provide context.

- **`rustc --print=native-static-libs ...`** (same as today)

    ```console
    note: Link against the following native artifacts when linking against this static library. The order and any duplication can be significant on some platforms.

    note: native-static-libs: -lglib-2.0 -lbar_cli -lgcc_s -lutil -lrt -lpthread -lm -ldl -lc -lsystemd -lfoo_cli
   ```

- **`rustc --print=native-static-libs=./native-static-libs.txt ...`**

    ```console
    note: Native artifacts to link against have been written to ./native-static-libs.txt. The order and any duplication can be significant on some platforms.
    ```

    native-static-libs.txt:

    ```txt
    -lglib-2.0 -lbar_cli -lgcc_s -lutil -lrt -lpthread -lm -ldl -lc -lsystemd -lfoo_cli
    ```

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 3]
```diff
@@ -1467,7 +1467,12 @@ fn print_native_static_libs(
     }
 
     match out {
-        OutFileName::Real(_) => out.overwrite(&lib_args.join(" "), sess),
+        OutFileName::Real(path) => {
+            out.overwrite(&lib_args.join(" "), sess);
```
Comment (b-naber) [inline compiler/rustc_codegen_ssa/src/back/link.rs:1471]: Can you move this into the if branch? Also I'm not sure what exact form `lib_args` has and whether it's straightforward to find the native artifacts in the file after reading the note on stdout, would it maybe be better to write this to the file: `"native-static-libs: {}", &lib_args.join(" ")`?

Comment (b-naber) [reply compiler/rustc_codegen_ssa/src/back/link.rs:1471]: ah nvm, I don't think this is necessary.

**Comment by b-naber** [comment]
Can you maybe open a PR to update https://doc.rust-lang.org/rustc/command-line-arguments.html after this is merged, please?

@bors r+ rollup

**Comment by matthiaskrgr** [comment]
@bors r- rollup=maybe
failed in a rollup https://github.com/rust-lang-ci/rust/actions/runs/5612166818/job/15205301430#step:24:49674

**Comment by Urgau** [comment]
Not a compiler member, but shouldn't this have gone through a compiler FCP since it's insta-stable or be gated as an unstable feature?

**Comment by apiraino** [comment]
For the record, this was discussed in a T-compiler meeting [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-07-27/near/379199602) (suggested to stick to the protocol and go througn an MCP/FCP process before stabilizing a user-facing change).

@rustbot label -I-compiler-nominated

**Pull request merged**